proposals/mutate_existing_resources.md

```diff
@@ -0,0 +1,216 @@
+# Kyverno Design Proposal - Mutate Existing Resources
+
+Created: March 21st, 2022
+Author: [Shuting Zhao](https://github.com/realshuting)
+
+
+Table of contents:
+  * [Overview](#overview)
+  * [Proposal](#proposal)
+  * [Examples](#examples)
+  * [Implementation](#implementation)
+    + [Watch Trigger Resources](#watch-trigger-resources)
+      - [1. Admission Webhook](#1-admission-webhook)
+      - [2. Event Watcher](#2-event-watcher)
+    + [Apply Policies](#apply-policies)
+    + [Mutate Existing Resources](#mutate-existing-resources)
+
+<small><i><a href='http://ecotrust-canada.github.io/markdown-toc/'>Table of contents generated with markdown-toc</a></i></small>
+
+
+## Overview
+
+Kyverno webhook controller runs as an admission controller in Kubernetes, and it mutates resources upon admission review. Such mutation is performed based on resources operations (CREATE/UPDATE/DELETE) via webhook.
+
+As of release 1.6.1, these are still the outstanding issues which Kyverno needs to support hence this proposal:
+
+- Mutate target resource which is different from the trigger resource, [#2139](https://github.com/kyverno/kyverno/issues/2139), [#1722](https://github.com/kyverno/kyverno/issues/1722).
+- Mutate trigger resources based on policy update, [#1607](https://github.com/kyverno/kyverno/issues/1607).
+
+*The **trigger** resource is matched in the policy while the **target** resource is the resource to be mutated*.
+
+
+## Proposal
+
+To support mutate existing resources:
+
+- `spec.rules.match` will be used to identify operations on the trigger resources
+    - To *"mutate trigger resources based on policy update"*, the syntax is slightly different in order to simplify policy definition, see example 3 below.
+- A new attribute `spec.rules.mutate.targets` will be introduced to specify a list of target resources to be mutated.
+- Kyverno will fetch and mutate the existing resources once the trigger resource changes, the user needs to grant Kyverno permissions to operate on target resources.
+- The patches will be added as an annotation to existing resources.
+- CLI support with the apply command, similar to background generation, to allow users to understand the impact such a policy may have on existing resources.
+
+
+
+
+## Examples
+
+1. This example adds the label `foo=bar` to deployment `staging/example-A` on configmap `staging/dictionary` update.
+
+```yaml
+apiVersion: kyverno.io/v1
+kind: ClusterPolicy
+metadata:
+  name: "..."
+spec:
+  rules:
+    - name: "..."
+      match:
```

Comment:
Since we're deprecating non `any`|`all` blocks under `match` and `exclude` may want to add this to ensure it's tested. Since we have to hang onto them for 1.7.0, it should work either way. Minor point but we've seen things break recently.

Reply:
Updated!